ykcompile/src/lib.rs

```diff
@@ -935,128 +778,65 @@ impl<TT> TraceCompiler<TT> {
-                            ; add Rq(reg), c_val as u32 as i32
-                        );
-                    } else {
-                        dynasm!(self.asm
-                            ; mov rax, QWORD c_val
-                            ; add Rq(reg), rax
-                        );
-                    }
-                }
-                Location::Mem(ro) => {
-                    if c_val <= u32::MAX.into() {
-                        dynasm!(self.asm
-                            ; add QWORD [Rq(ro.reg) + ro.offs], c_val as u32 as i32
-                        );
-                    } else {
-                        dynasm!(self.asm
-                            ; mov rax, QWORD c_val
-                            ; add [Rq(ro.reg) + ro.offs], rax
-                        );
-                    }
-                }
-                _ => todo!(),
-            },
-            1 => match l {
-                Location::Register(reg) => {
-                    dynasm!(self.asm
-                        ; add Rb(reg), c_val as u8 as i8
-                    );
-                }
-                Location::Mem(ro) => {
-                    dynasm!(self.asm
-                        ; add BYTE [Rq(ro.reg) + ro.offs], c_val as u8 as i8
-                    );
-                }
-                _ => todo!(),
-            },
-            _ => todo!("{}", size),
+        if c_val <= u32::MAX.into() {
+            dynasm!(self.asm
+                ; add Rq(l), c_val as u32 as i32
+            );
+        } else {
+            dynasm!(self.asm
+                ; mov rax, QWORD c_val
+                ; add Rq(l), rax
+            );
         }
     }
 
     /// Compile a TIR statement.
     fn c_statement(&mut self, stmt: &Statement) -> Result<(), CompileError> {
         match stmt {
             Statement::Assign(l, r) => {
+                let temp = self.create_temporary();
                 match r {
-                    Rvalue::Use(Operand::Place(p)) => {
-                        self.mov_place_place(l, p);
-                    }
+                    Rvalue::Use(Operand::Place(p)) => self.c_place(temp, p),
                     Rvalue::Use(Operand::Constant(c)) => match c {
-                        Constant::Int(ci) => self.mov_place_constint(l, ci),
-                        Constant::Bool(b) => self.mov_place_bool(l, *b),
+                        Constant::Int(ci) => self.c_constint(temp, ci),
+                        Constant::Bool(b) => self.c_bool(temp, *b),
                         c => todo!("{}", c),
                     },
                     Rvalue::CheckedBinaryOp(binop, op1, op2) => {
-                        self.c_checked_binop(l, binop, op1, op2)
+                        self.c_checked_binop(temp, binop, op1, op2);
+                        // FIXME deal with overflow
+                        let mut l = l.clone();
+                        l.projection.push(Projection::Field(0));
+                        // FIXME dedup code
```

Comment:
We should fix this now.

Reply:
Okay, I think it's not a super trivial fix though.

Reply:
I've changed this line in my branch, so perhaps let's leave it for this PR.